### PR TITLE
JaCoCo based code coverage analysis

### PR DESCRIPTION
**Pull request by rsvoboda**
JaCoCo based code coverage analysis

To get aggregated report I had to go via maven-antrun-plugin. I saw similar approach in other projects too.

Procedure example:
```
mvn -Pjacoco clean verify
mvn -Pjacoco-report verify -DskipTests -Denforcer.skip=true -Dcheckstyle.skip=true -Dskip.gradle.build=true
open target/jacoco-report/index.html

```

Report example: [jacoco-report.zip](https://github.com/quarkusio/quarkus/files/3018884/jacoco-report.zip)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -170,6 +171,125 @@
...
+                                        <report>
+                                            <executiondata>
+                                                <fileset dir="${basedir}">
+                                                    <include name="**/target/jacoco.exec"/>
+                                                </fileset>
+                                            </executiondata>
+                                            <structure name="JaCoCo for Quarkus">
+                                                <classfiles>
+                                                    <fileset dir="${basedir}">
+                                                        <include name="**/*999-SNAPSHOT.jar"/>
```
Comment (cescoffier) [inline pom.xml]: Can you use ${project.version} instead?

File: pom.xml [hunk 2]
```diff
@@ -170,6 +171,125 @@
...
+                                                        <exclude name="**/*quarkus-cli-999-SNAPSHOT.jar"/>
```
Comment (cescoffier) [inline pom.xml]: Any reason to exclude this one?

File: pom.xml [hunk 3]
```diff
@@ -170,6 +171,125 @@
                 </plugins>
             </build>
         </profile>
+        <profile>
+            <id>jacoco</id>
```
Comment (cescoffier) [inline pom.xml]: Can you add as a comment the instruction to run these?

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 2 above)
Comment (rsvoboda) [reply pom.xml]: JaCoCo doesn't like duplicated classes in and it throws 
`[ERROR] around Ant part ...<report>... @ 8:11 in /Users/rsvoboda/git/quarkus/target/antrun/build-main.xml: Error while analyzing devtools/aesh/target/original-quarkus-cli-999-SNAPSHOT.jar@META-INF/versions/11/org/wildfly/common/archive/JDKSpecific.class. Can't add different class with same name: org/wildfly/common/archive/JDKSpecific`

**Comment by rsvoboda** [comment]
@cescoffier PR updated

**Review by stuartwdouglas — APPROVED** [review]

**Comment by rsvoboda** [comment]
Rebased and added commit to this PR to:

 - have faster report generation command
 - avoid instrumenting test classes (io.quarkus.it.*)
 - have more generic report excludes

**Comment by Sanne** [comment]
@cescoffier your "requests for changes" is blocking this PR. I guess it's out of date?

**Pull request merged**